crates/runtime/src/vmcontext.rs

```diff
@@ -248,6 +256,28 @@ impl VMMemoryDefinition {
             current_length: other.current_length().into(),
         }
     }
+
+    /// In the configurations where bounds checks were elided in JIT code (because
+    /// we are using static memories with virtual memory guard pages) this manual
+    /// check is here so we don't segfault from Rust. For other configurations,
+    /// these checks are required anyways.
+    pub fn validate_addr(
+        &self,
+        addr: u64,
+        access_size: u64,
+        access_alignment: u64,
+    ) -> Result<*const u8, TrapCode> {
+        debug_assert!(access_alignment.is_power_of_two());
+        ensure!(addr % access_alignment == 0, TrapCode::HeapMisaligned);
+
+        let length = u64::try_from(self.current_length()).unwrap();
+        ensure!(
+            addr.saturating_add(access_size) < length,
+            TrapCode::HeapOutOfBounds
+        );
+        // SAFETY: checked above that the address is in bounds
+        Ok(unsafe { self.base.add(addr as usize) })
+    }
```

Comment:
I think it's ok to leave this scoped to `SharedMemory` for now since that's all it's used for

Reply:
See the comment on `unchecked_atomic_notify`.